db/flush_job.cc

```diff
@@ -630,6 +630,117 @@ bool FlushJob::MemPurgeDecider() {
+                           "Memtable Decider: ParseInternalKey did not parse "
+                           "entry_slice %s"
+                           "successfully.",
+                           entry_slice.data());
+          }
+          LookupKey lkey(res.user_key, kMaxSequenceNumber);
+          std::string vget;
+          Status s;
+          MergeContext merge_context;
+          SequenceNumber max_covering_tombstone_seq = 0, sqno = 0;
+
+          // Pick the oldest existing snapshot that is more recent
+          // than the sequence number of the sampled entry.
+          SequenceNumber min_seqno_snapshot = kMaxSequenceNumber;
+          SnapshotImpl min_snapshot;
+          for (SequenceNumber seq_num : existing_snapshots_) {
+            if (seq_num > res.sequence && seq_num < min_seqno_snapshot) {
+              min_seqno_snapshot = seq_num;
+            }
+          }
+          min_snapshot.number_ = min_seqno_snapshot;
+          ro.snapshot =
+              min_seqno_snapshot < kMaxSequenceNumber ? &min_snapshot : nullptr;
+
+          // Estimate if the sample entry is valid or not.
+          bool gres = mt->Get(lkey, &vget, nullptr, &s, &merge_context,
+                              &max_covering_tombstone_seq, &sqno, ro);
+          if (!gres) {
+            ROCKS_LOG_WARN(
+                db_options_.info_log,
+                "Memtable Get returned false when Get(sampled entry). "
+                "Yet each sample entry should exist somewhere in the memtable, "
+                "unrelated to whether it has been deleted or not.");
+          }
+          payload += entry_slice.size();
+
+          // TODO(bjlemaire): evaluate typeMerge.
+          // This is where the sampled entry is estimated to be
+          // garbage or not. Note that this is a garbage *estimation*
+          // because we do not include certain items such as
+          // CompactionFitlers triggered at flush, or if the same delete
+          // has been inserted twice or more in the memtable.
+          if (res.type == kTypeValue && gres && s.ok() &&
+              sqno == res.sequence) {
+            useful_payload += entry_slice.size();
+          } else if (((res.type == kTypeDeletion) ||
+                      (res.type == kTypeSingleDeletion)) &&
+                     s.IsNotFound() && gres) {
+            useful_payload += entry_slice.size();
+          }
+        }
+        if (payload > 0) {
+          estimated_useful_payload += useful_payload * 1.0 / payload;
+        } else {
+          ROCKS_LOG_WARN(
+              db_options_.info_log,
+              "Mempurge kSampling policy: null payload measured, and collected "
+              "sample size is %zu\n.",
+              sentries.size());
+        }
+        ROCKS_LOG_INFO(
+            db_options_.info_log,
+            "Mempurge kSampling policy: garbage ratio from sampling: %f.\n",
+            (payload - useful_payload) * 1.0 / payload);
```

Comment:
Potential divide by 0, which could be UB https://stackoverflow.com/questions/42926763/the-behaviour-of-floating-point-division-by-zero